R/shoji-catalog.R

```diff
@@ -240,28 +240,114 @@ setMethod("emails", "ShojiCatalog", function (x) getIndexSlot(x, "email"))
+    ...) {
+
+    index <- lapply(index(x), function(a) a[keys])
+    ## Return an empty dataframe if the function is called on an empty catalog
+    if (length(index) == 0) {
+        message("Catalog is empty, returning an empty dataframe")
+        return(data.frame())
+    } else {
+        ### The following code is equivalent to out <- purrr::map_df(index, entry_to_df)
+        ################
+        entry_list <- lapply(index, entryToDF, list_col_names = list_columns)
+        names   <- unique(unlist( lapply(entry_list, names)))
+        out <- data.frame(matrix(nrow = length(entry_list), ncol = length(names)))
+        names(out) <- names
+
+        for (i in seq_along(entry_list)) {
+            for (j in  names(entry_list[[i]])) {
+                out[i, j] <- entry_list[[i]][1, j]
+            }
+        }
+        #################
+
+        out <- as.data.frame(out, rownames = rownames, ...)
+
+        #When a bad key argument is passed to the the shoji index it returns a dataframe
+        # with an NA value. This exclude those columns.
+        exclude_cols <- grepl("^NA", names(out)) &
+            vapply(out, function(x)all(is.na(x)), FUN.VALUE = logical(1))
+        out <- out[, !exclude_cols, drop = FALSE]
+
+        missing_keys <- dQuote(keys[!(keys %in% names(out))])
+        if (any(!(keys %in% names(out))) && keys != TRUE) {
+            if (length(missing_keys) == 1) {
+                error_text <-  " is an invalid key for catalogs of class "
+            } else {
+                error_text <- " are invalid keys for catalogs of class "
+            }
+            halt(
+                serialPaste(missing_keys),
+                error_text,
+                class(x),
+                ".")
+        }
+
+        #reorder columns to match the order in which keys were supplied
+        out <- out[, keys, drop = FALSE]
+        return(out)
     }
-    return(out)
+}
+
+#' entryToDF
+#' Turns an entry in the catalog into a one-row dataframe
+#'
+#' @param entry
+#' A single entry in a catalog
+#'
+#' @param list_col_names
+#' Some entries in a catalog do not fit neatly into a traditional dataframe and are
+#' instead stored in a list-column. This is controlled by passing the attribute names
+#' as the `list_col_names` argument. `
+#'
+#' @return A one row data frame
+#'
+entryToDF <- function(entry, list_col_names ){
+            entry[vapply(entry, is.null, logical(1))] <- NA
+            vect_col <- entry[!(names(entry) %in% list_col_names)]
+            mislabled_list_cols <- lapply(vect_col, length) > 1
+
+            error_text <- " contain more than one entry and is not included in list_col_names"
+            if (sum(mislabled_list_cols) == 1) {
+                error_text <- " contains more than one entry and are not included in list_col_names"
+            }
+            if (any(mislabled_list_cols)) {
+                halt( serialPaste( dQuote(names(vect_col[mislabled_list_cols]))),
```

Comment:
You can remove these spaces here. We don't usually store things to be used once, but given how long this is, it might be worth storing `names(vect_col[mislabled_list_cols])` or `dQuote(names(vect_col[mislabled_list_cols]))` temporarily for readability (thought it's right on the border, so don't feel obligated).